app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -113,12 +113,28 @@ def save_child_inventory(obj, hashes, child_keys, *args)
 
   def store_ids_for_new_records(records, hashes, keys)
     keys = Array(keys)
-    hashes.each do |h|
-      r = records.detect { |r| keys.all? { |k| r.send(k) == r.class.type_for_attribute(k.to_s).cast(h[k]) } }
-      h[:id] = r.id
+    # Lets first index the hashes based on keys, so we can do O(1) lookups
+    record_index = {}
+    records.each do |record|
+      record_index[build_index_from_record(keys, record)] = record
+    end
+
+    record_class = records.first.class
+
+    hashes.each do |hash|
+      record = record_index[build_index_from_hash(keys, hash, record_class)]
+      hash[:id] = record.id
```

Comment:
@Fryguy ah, it's because save_child_inventory is called before we call association.push, so in a case of child_inventory for a newly created record, we are sending there record that was only built and not yet created. Therefore the record.id result in nil and this relation is only filled in a folloup refresh, where we update the existing record

So this seems like older and unrelated issue. :-)